PEP 622: Include overview to different patterns and summarised syntax (#1501)
python/peps: pep-0622.rst

Content-Type: text/x-rst
Created: 23-Jun-2020
Python-Version: 3.10

> Post-History: 8-Jul-2020

Reviewer: ```suggestion
Post-History: 23-Jun-2020, 8-Jul-2020
```

and the second one into ``b``.

This PEP defines several types of patterns. These are certainly not the
only possible ones, so the design decision was made to choose a subset of
functionality that is useful now but conservative. More patterns can be added
later as this feature gets more widespread use. See the `rejected ideas`_
and `deferred ideas`_ sections for more details.

The patterns listed here are described in more detail below, but summarized 
together in this section for simplicity:

- A **capture pattern** looks like ``x`` and is equivalent to an identical
  assignment target: it always matches and binds the variable
  with the given name.
- A **sequence pattern** looks like ``[a, *rest, b]`` and is similar to 
  a list unpacking. An important difference is that the elements nested
  within it can be any kind of patterns patterns, not just names or sequences.
  It matches only sequences of appropriate length, as long as all the sub-patterns
  also match. It makes all the bindings of its sub-patterns.
- The **wildcard pattern** is a single underscore: ``_``. It always matches,
  but does not capture any variable (which prevents interference with other

>   uses that you could have for ``_`` and allows some optimizations). 

Reviewer: This is a good addition, and now I realize the wildcard pattern isn't listed separately later -- it's hidden in the capture pattern. Also the order patterns are presented here is different from how they are presented later. I'd like this to be consistent.

Reviewer: Also maybe wildcard should immediately follow capture? And capture should mention `_` is an exception.

Author: I choose this order here to highlight the capture over the constant/literals. I understand that the order later on is design to build the syntax bottom up.

- A **mapping pattern** looks like ``{"user": u, "emails": [*es]}``. It matches
  mappings with at least the set of provided keys, and if all the
  sub-patterns match their corresponding values. It binds whatever the
  sub-patterns bind while matching with the values corresponding to the keys.
  Usage of ``**`` to capture extra keys is allowed.
- A **class pattern** is similar to the above but matches attributes instead
  of keys. It looks like ``datetime.date(year=y, day=d)``. It matches
  instances of the given type, having at least the specified
  attributes, as long as the attributes match with the corresponding
  sub-patterns. It binds whatever the sub-patterns bind when matching with the
  values of
  the given attributes. An optional protocol also allows matching positional
  arguments.
- A **literal pattern** is useful to filter constant values in a structure.

>   It looks like a python literal (including some python singletons like 

Reviewer: ```suggestion
  It looks like a Python literal (including some Python singletons like 
```

>   It looks like a python literal (including some python singletons like 

Reviewer 2: ```suggestion
  It looks like a Python literal (including values like ``True``, ``False``, and
```

  ``None``). It only matches objects equal to the literal, and
  never binds.
- A **constant value pattern** works like the above for certain named

>   constants. Note that there are syntax restrictions given the possible

Reviewer 2: ```suggestion
  constants. Note that it must be a qualified (dotted) name, given the possible
```

  ambiguity with a capture pattern. It may look like ``Color.RED`` and
  only matches values equal to the corresponding value. It never binds.

> - An **or pattern** looks like ``[*x] | {"elems": [*x]}``. It matches if any

Reviewer: ```suggestion
- An **OR pattern** looks like ``[*x] | {"elems": [*x]}``. It matches if any
```

>   of its sub-patterns match. It uses the binding for the leftmost match attempt

Reviewer 2: ```suggestion
  of its sub-patterns match. It uses the binding for the leftmost pattern
```

Literal Patterns
~~~~~~~~~~~~~~~~

Simplified syntax::

    literal_pattern:

>         | signed_number

Reviewer: If we're simplifying, maybe just "number"? And "string" below?

Constant Value Patterns
~~~~~~~~~~~~~~~~~~~~~~~

Simplified syntax::

>     constant_pattern: NAME ('.' NAME)*

Reviewer: ```suggestion
    constant_pattern: NAME ('.' NAME)+
```

Constant Value Patterns
~~~~~~~~~~~~~~~~~~~~~~~

Simplified syntax::

> constant_pattern: NAME ('.' NAME)*

Reviewer 3: Should be indented?

Author: yes, just fixed this, thanks

Sequence Patterns
~~~~~~~~~~~~~~~~~

Simplified syntax::

  sequence_pattern: '[' [values_pattern] ']' | '(' [values_pattern] ')'
  values_pattern: ','.value_pattern+ ','?

>   value_pattern: '*' capture_pattern | pattern

Reviewer: There's a lie here -- `(foo)` is not a sequence pattern, but `[foo]` and `(foo,)` are. Not sure how to fix without adding another helper rule (like in the Appendix).

Reviewer 2: `'(' [value_pattern ',' [values pattern]] ')'`?

Reviewer 2: Ah, should we add another alt for parenthesized tuples?

`'[' [values_pattern] ']' | '(' [value_pattern ',' [values pattern]] ')' | value_pattern ',' [values pattern]`

Reviewer: I would leave that out here. In the real grammar it's handled differently. If we draw more attention to it here we'll just get complaints about ambiguity in `MyClass(p1, p2)`. Maybe a note in the text is better?

Author: I'll add a note on the first simplified grammar saying "this is for readability, go to the appendix for real specs". In nicer words.

Simplified syntax::

  sequence_pattern:
      | '[' [values_pattern] ']' 
      | '(' [value_pattern ',' [values pattern]] ')' 

>       | value_pattern ',' [values pattern]

Reviewer: I'd drop this line (see previous exchange).

Author: dropped and added a note

Mapping Patterns
~~~~~~~~~~~~~~~~

Simplified syntax::

>   mapping_pattern: '{' items_pattern? '}'

Reviewer: ```suggestion
  mapping_pattern: '{' [items_pattern] '}'
```

     - Why we choose ``_`` for wildcard patterns
     - Why we choose ``|`` for OR patterns
     - Why we choose not to use special syntax for capture variables
     - Why this pattern matching operation and not others

   - Clarify exception and side effect semantics
   - Clarify partial binding semantics
   - Drop restriction on use of ``_`` in load contexts
   - Simplify behavior of ``__match_args__``
   - Drop the ``__match__`` protocol (moved to `deferred ideas`_)
   - Drop ``ImpossibleMatchError`` exception
   - Drop leading dot for loads (moved to `deferred ideas`_)
   - Reworked the initial sections (everything before `syntax`_)
   - Added an overview of all the types of patterns before the 

>      detailed description.

Reviewer: ```suggestion
     detailed description
```
